frame/contracts/src/wasm/runtime.rs

```diff
@@ -799,17 +807,19 @@ define_env!(Env, <E: Ext>,
 		Ok(())
 	},
 
-	// Record a request to restore the caller contract to the specified contract.
+	// Try to restore the given destination contract sacrificing the caller.
 	//
-	// At the finalization stage, i.e. when all changes from the extrinsic that invoked this
-	// contract are committed, this function will compute a tombstone hash from the caller's
-	// storage and the given code hash and if the hash matches the hash found in the tombstone at
-	// the specified address - kill the caller contract and restore the destination contract and set
-	// the specified `rent_allowance`. All caller's funds are transferred to the destination.
+	// This function will compute a tombstone hash from the caller's storage and the given code hash
+	// and if the hash matches the hash found in the tombstone at the specified address - kill
+	// the caller contract and restore the destination contract and set the specified `rent_allowance`.
+	// All caller's funds are transfered to the destination.
 	//
-	// This function doesn't perform restoration right away but defers it to the end of the
-	// transaction. If there is no tombstone in the destination address or if the hashes don't match
-	// then restoration is cancelled and no changes are made.
+	// If there is no tombstone at the destination address or if the hashes don't match
+	// then restoration is cancelled and no changes are made. Additionally, this contract instance
+	// should be not present on the contract call stack. If any of these preconditions are not held,
+	// then a trap is generated.
+	//
+	// Otherwise, the contract execution is terminated and the destination contract is restored.
```

Comment:
I feel that it does not properly express that this function is diverging for any argument.

```suggestion
	// If there is no tombstone at the destination address, the hashes don't match or this contract
	// instance is already present on the contract call stack, a trap is generated.
	//
	// Otherwise, the destination contract is restored. This function is diverging and stops execution
	// even on success.
```